File: app/src/main/java/com/melvinhou/dimension2/media/picture/PictureActivity.java
package com.melvinhou.dimension2.media.picture;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.bumptech.glide.request.RequestOptions;
import com.melvinhou.dimension2.R;
import com.melvinhou.dimension2.databinding.ActPictureBD;
import com.melvinhou.kami.mvvm.DataBindingActivity;
import com.melvinhou.kami.util.DimenUtils;
import com.melvinhou.kami.wiget.PhotoCutterView;

/**
 * ===============================================
 * = 作 者：风 尘
 * <p>
 * = 版 权 所 有：melvinhou@163.com
 * <p>
 * = 地 点：中 国 北 京 市 朝 阳 区
 * <p>
 * = 时 间：2020/7/10 3:50
 * <p>
 * = 分 类 说 明：图片查看
 * ================================================
 */
public class PictureActivity extends DataBindingActivity<ActPictureBD> {

    //文件查找格式
    public static final String FILE_UNSPECIFIED = "*/*";
    //图片查找格式
    public static final String IMAGE_UNSPECIFIED = "image/*";
    //文件
    public static final int FILE_REQUEST_CODE = 0;
    //相册
    public static final int ALBUM_REQUEST_CODE = 1;
    //相机
    public static final int CAMERA_REQUEST_CODE = 2;

    private String url;

    @Override
    protected void initWindowUI() {
//        super.initWindowUI();
        getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN |
                        View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR |
                        View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR |
                        View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        int statusColor = 0x40ffffff;
        getWindow().setStatusBarColor(statusColor);
        getWindow().setNavigationBarColor(statusColor);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount() == 0) {
            event.startTracking();//这个会执行过渡动画
            return true;
        }
        return false;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                //finish()不会执行动画
                finishAfterTransition();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected int getLayoutID() {
        return R.layout.activity_picture;
    }

    @Override
    protected void initView() {
    }

    @Override
    protected void initListener() {
        getViewDataBinding().tools.getChildAt(0).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                updatePhoto();
            }
        });
    }

    @Override
    protected void initData() {
        url = getIntent().getStringExtra("url");
        RequestOptions options = new RequestOptions();
//        Glide.with(FcUtils.getContext()).asBitmap().load(url).apply(options).into(getViewDataBinding().photo);

        getViewDataBinding().setUrl(url);
        //图片查看模式
        int mode = getIntent().getIntExtra("mode", PhotoCutterView.GESTURE_MODE_INFINITE);
        getViewDataBinding().photo.setGestureMode(mode);
        if (mode == PhotoCutterView.GESTURE_MODE_BOX) {
            getViewDataBinding().photo.setCenterCheckBox(
                    getIntent().getIntExtra("boxSize",DimenUtils.dp2px(300)));
            getViewDataBinding().photo.setCheckBoxColor(
                    getIntent().getIntExtra("boxColor",0x80000000));
        }
    }

    @Override
    protected void onLoading() {
//        super.onLoading();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.picture, menu);
        return super.onCreateOptionsMenu(menu);
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        Log.e("onActivityResult", "requestCode=" + requestCode);
        if (resultCode == Activity.RESULT_OK) {//成功加载
            if (requestCode == 0 || requestCode == 1 || requestCode == 2) {
                updatePhoto(data.getData());
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    /**
     *
     */
    public void updatePhoto() {
        Intent intent = new Intent();
        // 开启Pictures画面Type设定为image
        intent.setType(IMAGE_UNSPECIFIED);
        // 使用Intent.ACTION_GET_CONTENT这个Action
        intent.setAction(Intent.ACTION_GET_CONTENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        // 取得相片后返回本画面
        startActivityForResult(intent, ALBUM_REQUEST_CODE);
    }


    private void updatePhoto(Uri uri) {

        getViewDataBinding().photo.setImageURI(uri);
    }

}

File: media-sample/src/main/java/com/melvinhou/media_sample/photo/PictureActivity.kt
package com.melvinhou.media_sample.photo

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.text.InputType
import android.text.TextUtils
import android.util.Log
import android.view.MenuItem
import androidx.core.view.isVisible
import com.melvinhou.kami.io.DownloadHelper
import com.melvinhou.kami.mvvm.BaseViewModel
import com.melvinhou.kami.mvvm.BindActivity
import com.melvinhou.kami.util.DimenUtils
import com.melvinhou.kami.util.FcUtils
import com.melvinhou.kami.view.wiget.PhotoCutterView
import com.melvinhou.knight.KUITools
import com.melvinhou.knight.loadImage
import com.melvinhou.media_sample.R
import com.melvinhou.media_sample.databinding.FragmentPictureBinding


/**
 * ===============================================
 * = 作 者：风 尘
 * <p>
 * = 版 权 所 有：melvinhou@163.com
 * <p>
 * = 地 点：中 国 北 京 市 朝 阳 区
 * <p>
 * = 时 间：2023/4/23 0023 14:21
 * <p>
 * = 分 类 说 明：图片浏览
 * ================================================
 */
class PictureActivity : BindActivity<FragmentPictureBinding, BaseViewModel>() {

    override fun openViewBinding(): FragmentPictureBinding =
        FragmentPictureBinding.inflate(layoutInflater)

    override fun openModelClazz(): Class<BaseViewModel> =
        BaseViewModel::class.java


    private val TAG = PictureActivity::class.java.name
    private var mPath: String? = null

    override fun upBarMenuID(): Int = R.menu.bar_photo

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_save -> {
                savePicture()
                return true
            }
            R.id.menu_open -> {
                open()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun backward() {
        //finish()不会执行动画
        finishAfterTransition()
    }

    override fun initView() {
        mBinding.barRoot.title.text = "图片详情"
        mBinding.barRoot.barRoot.setBackgroundColor(0x80FFFFFF.toInt())

        //图片查看模式
        val mode = intent.getIntExtra("mode", PhotoCutterView.GESTURE_MODE_INFINITE)
        mBinding.container.setGestureMode(mode)
        if (mode == PhotoCutterView.GESTURE_MODE_BOX) {
            mBinding.container.setCenterCheckBox(
                intent.getIntExtra("boxSize", DimenUtils.dp2px(300)).toFloat()
            )
            mBinding.container.setCheckBoxColor(
                intent.getIntExtra("boxColor", 0x80000000.toInt())
            )
        }
    }

    override fun initListener() {
        mBinding.container.setOnClickListener {
            //标题栏
            val isVisible = mBinding.barRoot.barRoot.isVisible
            mBinding.barRoot.barRoot.isVisible = !isVisible
        }
    }

    override fun initData() {
        val url = intent.getStringExtra("url")
        if (TextUtils.isEmpty(url)) {
            KUITools.showInputDialog01(
                this,
                "图片地址",
                "输入需要打开的图片url",
                InputType.TYPE_CLASS_TEXT
            ) {
                loadData(it)
            }
        } else {
            loadData(url)
        }
    }

    private fun loadData(url: String?) {
        mPath = url;
        mBinding.container.loadImage(url, -1, -1)
    }

    /**
     * 保存图片
     */
    private fun savePicture() {
        mBinding.container.apply {
            if (drawable == null) return
            if (gestureMode == PhotoCutterView.GESTURE_MODE_BOX) {//裁剪
                val rect = checkBoxMargin
                val size = checkBoxSize
                val bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.RGB_565)
                val canvas = Canvas(bitmap)
                canvas.translate(-rect[0], -rect[1])
                draw(canvas)
//                Bitmap.createBitmap(bitmap,rect[0],rect[1],size,size)//图片裁剪
                mBinding.ivShow.setImageBitmap(bitmap)
            } else {
//                val bitmap = Bitmap.createBitmap(
//                    drawable.intrinsicWidth,
//                    drawable.intrinsicHeight,
//                    Bitmap.Config.ARGB_8888
//                )
//                val canvas = Canvas(bitmap)
//                drawable.draw(canvas)
//                mBinding.ivShow.setImageBitmap(bitmap)
                mPath?.let {
//                    val name  = FileUtils.getFileNameForDate()+".jpg"
                    val name = it.split("/").last()
                    //下载
                    Log.w(TAG, "文件：${name},地址：${it}")
                    val helper = DownloadHelper.getInstance(name, it)
                    helper.setDownloadListener(object : DownloadHelper.DownloadListener {
                        override fun onStart() {
                        }

                        override fun onProgress(soFarSize: Long, totalSize: Long) {

                        }

                        override fun onFinish(fileFullPath: String?, totalSize: Long) {
                            FcUtils.showToast("保存成功：${fileFullPath}")
                        }

                        override fun onFailed() {
                            FcUtils.showToast("保存失败")
                        }
                    })
                    helper.start()
                }
            }
        }
    }

    private fun open() {
        val intent = Intent()
        // 开启Pictures画面Type设定为image
        intent.type = "image/*"
        // 使用Intent.ACTION_GET_CONTENT这个Action
        intent.action = Intent.ACTION_GET_CONTENT
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        toResultActivity(intent) {
            if (it.resultCode == RESULT_OK) { //成功加载
                loadData(it.data?.data.toString())
            }
        }
    }


}